### PR TITLE
[Model] Add support for EXAONE-4.0 Model

### PR DESCRIPTION
**Pull request by BoxBy**
## Motivation

Adding support for [LGAI](https://github.com/LG-AI-EXAONE)'s [EXAONE-4.0](https://github.com/LG-AI-EXAONE/EXAONE-4.0) 
The sliding window code was implemented by referencing [gemma2](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/gemma2.py)

## Modifications
Add exaone4.py

To use the model for reasoning, use the `--reasoning-parser deepseek-r1` argument.

**Example:**
```
python -m sglang.launch_server \
--model-path LGAI-EXAONE/EXAONE-4.0-32B \
... \
--reasoning-parser deepseek-r1 # for reasoning
```

## Comparison with ifeval score
Offiial ifeval score
[EXAONE-4.0-32B](https://huggingface.co/LGAI-EXAONE/EXAONE-4.0-32B): **83.7**(Reasoning), **84.8**(Non-Reasoning)
[EXAONE-4.0-1.2B](https://huggingface.co/LGAI-EXAONE/EXAONE-4.0-1.2B): **67.8**(Reasoning), **74.7**(Non-Reasoning)

The vLLM score was benchmarked using the code in [vllm-project/vllm#21060](https://github.com/vllm-project/vllm/pull/21060).

SGLang - EXAONE-4.0-1.2B

> |Tasks |Version|Filter|n-shot|        Metric         |   |Value |   |Stderr|
> |------|------:|------|-----:|-----------------------|---|-----:|---|------|
> |ifeval|      4|none  |     0|inst_level_loose_acc   |↑  |0.8393|±  |   N/A|
> |      |       |none  |     0|inst_level_strict_acc  |↑  |0.8153|±  |   N/A|
> |      |       |none  |     0|prompt_level_loose_acc |↑  |0.7708|±  |0.0181|
> |      |       |none  |     0|prompt_level_strict_acc|↑  |0.7431|±  |0.0188|

vLLM - EXAONE-4.0-1.2B

> |Tasks |Version|Filter|n-shot|        Metric         |   |Value |   |Stderr|
> |------|------:|------|-----:|-----------------------|---|-----:|---|------|
> |ifeval|      4|none  |     0|inst_level_loose_acc   |↑  |0.8513|±  |   N/A|
> |      |       |none  |     0|inst_level_strict_acc  |↑  |0.8297|±  |   N/A|
> |      |       |none  |     0|prompt_level_loose_acc |↑  |0.7856|±  |0.0177|
> |      |       |none  |     0|prompt_level_strict_acc|↑  |0.7560|±  |0.0185|

SGLang - EXAONE-4.0-1.2B(Reasoning)

> |Tasks |Version|Filter|n-shot|        Metric         |   |Value |   |Stderr|
> |------|------:|------|-----:|-----------------------|---|-----:|---|------|
> |ifeval|      4|none  |     0|inst_level_loose_acc   |↑  |0.8477|±  |   N/A|
> |      |       |none  |     0|inst_level_strict_acc  |↑  |0.8249|±  |   N/A|
> |      |       |none  |     0|prompt_level_loose_acc |↑  |0.7800|±  |0.0178|
> |      |       |none  |     0|prompt_level_strict_acc|↑  |0.7523|±  |0.0186|

vLLM - EXAONE-4.0-1.2B(Reasoning)

> |Tasks |Version|Filter|n-shot|        Metric         |   |Value |   |Stderr|
> |------|------:|------|-----:|-----------------------|---|-----:|---|------|
> |ifeval|      4|none  |     0|inst_level_loose_acc   |↑  |0.8513|±  |   N/A|
> |      |       |none  |     0|inst_level_strict_acc  |↑  |0.8345|±  |   N/A|
> |      |       |none  |     0|prompt_level_loose_acc |↑  |0.7856|±  |0.0177|
> |      |       |none  |     0|prompt_level_strict_acc|↑  |0.7671|±  |0.0182|

## Checklist

- [X] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [X] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [X] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/models/exaone4.py [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+        q = self.q_norm(q)
+        q = q.reshape(q_shape)
+
+        k_shape = k.shape
+        k = k.reshape(-1, self.head_dim)
+        k = self.k_norm(k)
+        k = k.reshape(k_shape)
+
+        if not self.sliding_window_pattern or self.is_sliding:
+            q, k = self.rotary_emb(positions, q, k)
```
[the remark this answers is not in the document]
Comment (BoxBy) [reply python/sglang/srt/models/exaone4.py:214]: Unlike other models, EXAONE-4.0 uses a unique hybrid attention structure. The conditional application of apply_rotary_pos_emb shown below is necessary for this architecture.


```python
# We use global NoPE for hybrid attention model
if self.sliding_window is None or self.is_sliding:
    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
```
You can find this logic at the [link you referenced](https://github.com/lgai-exaone/transformers/blob/add-exaone4/src/transformers/models/exaone4/modeling_exaone4.py#L251).

File: python/sglang/srt/models/exaone4.py [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+            if tp_size > 1 and ("o_proj" in name or "down_proj" in name):
+                gathered_weights = [torch.zeros_like(weight) for _ in range(tp_size)]
+                torch.distributed.all_gather(gathered_weights, weight)
+                weight = torch.cat(gathered_weights, dim=1)
+            return weight.cpu().to(torch.float32).numpy().tolist()[:truncate_size]
+
+        except Exception:
+            logger.error(
+                f"Error getting weights by name {name} in LlamaForCausalLM: {get_exception_traceback()}"
+            )
```
[the remark this answers is not in the document]
Comment (BoxBy) [reply python/sglang/srt/models/exaone4.py:686]: fixed at 5115bc1f28f4f23aa2cf0acdc29a6f729a1b0be0

File: python/sglang/srt/models/exaone4.py [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+logger = logging.getLogger(__name__)
+
+
+# Aligned with HF's implementation, using sliding window inclusive with the last token
+# SGLang assumes exclusive
+def get_attention_sliding_window_size(config):
+    if getattr(config, "sliding_window", None) is not None:
+        return config.sliding_window
+    else:
+        return None
```
[the remark this answers is not in the document]
Comment (BoxBy) [reply python/sglang/srt/models/exaone4.py:47]: Unlike Gemma2, EXAONE-4.0 has a different architecture, so we use the standard sliding window implementation directly for this model.

File: python/sglang/srt/models/exaone4.py [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+            )
+        if not self.pp_group.is_last_rank:
+            return PPProxyTensors(
+                {
+                    "hidden_states": hidden_states,
+                    "residual": residual,
+                }
+            )
+        else:
+            hidden_states, _ = self.norm(hidden_states, residual)
```
Comment (lkm2835) [inline python/sglang/srt/models/exaone4.py]: It seems that `residual` should be removed.
 
You can refer to [vllm PR](https://github.com/vllm-project/vllm/pull/21060).

File: python/sglang/srt/models/exaone4.py [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+
+        # Fully Connected
+        hidden_states = self.mlp(hidden_states)
+
+        # Use post-LN
+        hidden_states = self.post_feedforward_layernorm(hidden_states)
+        orig_dtype = hidden_states.dtype
+        hidden_states = hidden_states.to(torch.float32)
+        hidden_states = hidden_states + residual.to(torch.float32)
+        hidden_states = hidden_states.to(orig_dtype)
```
Comment (lkm2835) [inline python/sglang/srt/models/exaone4.py]: Is it necessary to upcast to `float32`?

**Comment by lkm2835** [comment]
Hi, @BoxBy. Thanks for supporting EXAONE-4.0 on SGLang.
I've found a few parts that seem to need some fixes.

**Comment by BoxBy** [comment]
> Hi, @BoxBy. Thanks for supporting EXAONE-4.0 on SGLang. I've found a few parts that seem to need some fixes.

Thanks for the feedback. All feedback addressed at 694a651632aabeaa4fc777ebb87e7ac1d3635b33
The float32 casting was there because it was part of the original vLLM [RMSNorm](https://github.com/Deepfocused/vllm-exaone4.0/blob/main/vllm/model_executor/layers/layernorm.py#L118) implementation, but it has now been removed.

**Comment by lkm2835** [comment]
What has been changed in `exaone.py`?

**Comment by BoxBy** [comment]
> What has been changed in `exaone.py`?

My apologies, the changes to `exaone.py` were included by mistake
I've reverted it back to the version from ada268fd05b214ca8c674517d6d46958ee476af8.

**Comment by lgai-exaone** [comment]
@zhyncs @ispobock @ByronHsu @zhaochenyang20
Could you kindly review this PR?

**Comment by lkm2835** [comment]
Hi @BoxBy,
Could you test with inputs longer than 4k to check if the `sliding window` is applied?
It also seems that options like `slidding window` as used in `gpt-oss` should be added.

https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2604
https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2643

**Comment by mickqian** [comment]
I noticed that the accuracy of these two models in SGLang is lower than in vLLM, any reasons?

**Comment by lkm2835** [comment]
https://github.com/sgl-project/sglang/pull/8205#discussion_r2218575379

If the accuracy was evaluated before modifying this `norm`, the results might be lower.

**Comment by BoxBy** [comment]
> Hi @BoxBy, Could you test with inputs longer than 4k to check if the `sliding window` is applied? It also seems that options like `slidding window` as used in `gpt-oss` should be added.
> 
> https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2604
> 
> 
> https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2643

Unfortunately, due to my graduate school graduation, I do not have any computing resources available to test this.
https://github.com/BoxBy/sglang/blob/add_exaone4/python/sglang/srt/models/exaone4.py#L165
The sliding_window code itself is implemented and can be activated by checking the `sliding_window_pattern` in the config.

**Comment by BoxBy** [comment]
> I noticed that the accuracy of these two models in SGLang is lower than in vLLM, any reasons?

As explained [here](https://github.com/sgl-project/sglang/pull/8205#issuecomment-3292632060)(thank you!), the results may differ because they were evaluated before modifying the `norm`.

**Review by mickqian — COMMENTED** [review]
Great work! Besides, you need to add copyright, change docs and test.

You can see [here](https://github.com/sgl-project/sglang/pull/8189#pullrequestreview-3225022123) for a reference, cheers

File: python/sglang/srt/models/exaone4.py [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+from sglang.srt.managers.schedule_batch import global_server_args_dict
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch, PPProxyTensors
+from sglang.srt.model_loader.weight_utils import (
+    default_weight_loader,
+    maybe_remap_kv_scale_name,
+)
+from sglang.srt.utils import add_prefix, make_layers
+from sglang.utils import get_exception_traceback
+
+logger = logging.getLogger(__name__)
```
Comment (mickqian) [inline python/sglang/srt/models/exaone4.py]: use sgl's logger

**Comment by lkm2835** [comment]
> > Hi @BoxBy, Could you test with inputs longer than 4k to check if the `sliding window` is applied? It also seems that options like `slidding window` as used in `gpt-oss` should be added.
> > https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2604
> > 
> > https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2643
> 
> Unfortunately, due to my graduate school graduation, I do not have any computing resources available to test this. https://github.com/BoxBy/sglang/blob/add_exaone4/python/sglang/srt/models/exaone4.py#L165 The sliding_window code itself is implemented and can be activated by checking the `sliding_window_pattern` in the config.

In SGLang, some attention backends don't support sliding window, it would be better to restrict to supported backends.

**Comment by mickqian** [comment]
> > > Hi @BoxBy, Could you test with inputs longer than 4k to check if the `sliding window` is applied? It also seems that options like `slidding window` as used in `gpt-oss` should be added.
> > > https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2604
> > > 
> > > https://github.com/sgl-project/sglang/blob/5afd0365334c113bdfb295687e353c4001f8f2e6/python/sglang/srt/server_args.py#L2643
> > 
> > 
> > Unfortunately, due to my graduate school graduation, I do not have any computing resources available to test this. https://github.com/BoxBy/sglang/blob/add_exaone4/python/sglang/srt/models/exaone4.py#L165 The sliding_window code itself is implemented and can be activated by checking the `sliding_window_pattern` in the config.
> 
> In SGLang, some attention backends don't support sliding window, it would be better to restrict to supported backends.

@lkm2835 You can check the server_args inside specific model classes

**Comment by BoxBy** [comment]
https://github.com/sgl-project/sglang/pull/8205#issuecomment-3294667740
https://github.com/sgl-project/sglang/pull/8205#discussion_r2349593785

Fixed it

File: python/sglang/srt/server_args.py [hunk 7]
```diff
@@ -2664,6 +2664,16 @@ def model_specific_adjustments(self):
                 f"Disable hybrid SWA memory for {model_arch} as it is not yet supported."
             )
             self.disable_hybrid_swa_memory = True
+        elif model_arch in ["Exaone4ForCausalLM"]:
+            if hf_config.sliding_window_pattern is not None:
+                # https://docs.sglang.ai/advanced_features/attention_backend.html
+                assert self.attention_backend in {
+                    "flashinfer",
```
Comment (lkm2835) [inline python/sglang/srt/server_args.py]:
```suggestion
```
Accuracy of the `flashinfer` attention backend is lower than that of the `triton` attention backend and vLLM

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -2664,6 +2664,16 @@ def model_specific_adjustments(self):
...
+                    "fa3",
+                    "triton",
+                    "trtllm_mha",
+                }, "flashinfer, fa3, triton, or trtllm_mla is required for Exaone4ForCausalLM-32B"
+
```
Comment (lkm2835) [inline python/sglang/srt/server_args.py]: Could you add `self.disable_hybrid_swa_memory = True`?

**Review by lkm2835 — CHANGES_REQUESTED** [review]

**Comment by BoxBy** [comment]
I apologize for the delay. https://github.com/sgl-project/sglang/pull/8205#discussion_r2352219436 https://github.com/sgl-project/sglang/pull/8205#discussion_r2352224776 Fixed

**Comment by BoxBy** [comment]
I apologize for the delay. https://github.com/sgl-project/sglang/pull/8205#discussion_r2352219436, https://github.com/sgl-project/sglang/pull/8205#discussion_r2352224776 fixed

**Comment by lkm2835** [comment]
Hi @BoxBy, could you resolve conflicts?

**Comment by BoxBy** [comment]
> Hi @BoxBy, could you resolve conflicts?

fixed

**Comment by lkm2835** [comment]
Hi @mickqian, could you help review this PR?
@BoxBy, could you also remove [flashinfer](https://github.com/sgl-project/sglang/pull/8205/files#diff-700b5118b493d60d7b5994857f5f1e6a7e842ad702392b8ab199945764dfc8edR703) from the assert log?

**Comment by BoxBy** [comment]
> Hi @mickqian, could you help review this PR? @BoxBy, could you also remove [flashinfer](https://github.com/sgl-project/sglang/pull/8205/files#diff-700b5118b493d60d7b5994857f5f1e6a7e842ad702392b8ab199945764dfc8edR703) from the assert log?

I apologize for the omission. It has been fixed.

**Review by lkm2835 — APPROVED** [review]

**Comment by lkm2835** [comment]
Hi @lifuhuang @mickqian, Could you review this PR to add support for the model?

**Comment by lkm2835** [comment]
Hi @zhyncs , could you help review this PR?

**Comment by JustinTong0323** [comment]
Sorry for accidentally closed it...
> ⚠️ Important: Chat Template Required
To use this model correctly, use the --chat-template argument to specify its corresponding chat template.
The required chat template file can be found in the [Hugging Face Repository](https://huggingface.co/collections/LGAI-EXAONE/exaone-40-686b2e0069800c835ed48375).

I am not quite understand it, why we should explicitly specify the chat-template? It would automatically applied by server if it is located in hf model file.

**Comment by lkm2835** [comment]
> I am not quite understand it, why we should explicitly specify the chat-template? It would automatically applied by server if it is located in hf model file.

I agree with that specifying the chat template doesn’t seem necessary.

**Comment by BoxBy** [comment]
> Sorry for accidentally closed it...
> 
> > ⚠️ Important: Chat Template Required
> > To use this model correctly, use the --chat-template argument to specify its corresponding chat template.
> > The required chat template file can be found in the [Hugging Face Repository](https://huggingface.co/collections/LGAI-EXAONE/exaone-40-686b2e0069800c835ed48375).
> 
> I am not quite understand it, why we should explicitly specify the chat-template? It would automatically applied by server if it is located in hf model file.

fixed
